### PR TITLE
Expose port number when the GPRC listener is initialized with port 0

### PR DESCRIPTION
**Pull request by edmondop**
**What changed?**

Solves #4329 

**Why?**

As of today, when using the Port 0 in the RPC config, a port is assigned from the operating system but this is lost along the way.

**How did you test it?**

- I run the CLI start-dev repointing the server directory to my local checkout
- I run make start locally

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

**Review by MichaelSnowden — COMMENTED** [review]
When I change my local config to use port 0, I get errors when starting the server here: 
image
image

**Comment by edmondop** [comment]
> When I change my local config to use port 0, I get errors when starting the server here: image image

I tried locally, the only error level problem I got was the following:

```
{"level":"error","ts":"2023-08-11T10:35:01.476-0700","msg":"Unable to process new range","shard-id":4,"address":"127.0.0.1:7234","component":"timer-queue-processor","error":"shard status unknown","logging-call-at":"queue_base.go:316","stacktrace":"go.temporal.io/server/common/log.(*zapLogger).Error\n\t/Users/edmondoporcu/Development/open-source/temporal/common/log/zap_logger.go:156\ngo.temporal.io/server/service/history/queues.(*queueBase).processNewRange\n\t/Users/edmondoporcu/Development/open-source/temporal/service/history/queues/queue_base.go:316\ngo.temporal.io/server/service/history/queues.(*scheduledQueue).processEventLoop\n\t/Users/edmondoporcu/Development/open-source/temporal/service/history/queues/queue_scheduled.go:218"}
{"level":"error","ts":"2023-08-11T10:35:01.476-0700","msg":"Unable to process new range","shard-id":4,"address":"127.0.0.1:7234","component":"archival-queue-processor","error":"shard status unknown","logging-call-at":"queue_base.go:316","stacktrace":"go.temporal.io/server/common/log.(*zapLogger).Error\n\t/Users/edmondoporcu/Development/open-source/temporal/common/log/zap_logger.go:156\ngo.temporal.io/server/service/history/queues.(*queueBase).processNewRange\n\t/Users/edmondoporcu/Development/open-source/temporal/service/history/queues/queue_base.go:316\ngo.temporal.io/server/service/history/queues.(*scheduledQueue).processEventLoop\n\t/Users/edmondoporcu/Development/open-source/temporal/service/history/queues/queue_scheduled.go:218"}
```

Can you please retry with a fresh docker-compose ?

File: common/config/config.go [hunk 1]
```diff
@@ -632,3 +633,14 @@ func (p *JWTKeyProvider) HasSourceURIsConfigured() bool {
...
+func (s *RPC) GetGRPCPort() int {
+	if s.boundGRPCPort != nil {
+		return *s.boundGRPCPort
+	}
+	return s.GRPCPort
+}
+
+func (r *RPC) SetBoundGRPCPort(port int) {
+	r.boundGRPCPort = &port
+}
```
[the remark this answers is not in the document]
Comment (edmondop) [reply common/config/config.go]: Fixed with https://github.com/temporalio/temporal/pull/4713/commits/0cc91e4fbd4a61784b6611663d56ea073bbbf005

File: common/config/config.go [hunk 2]
```diff
@@ -86,7 +86,8 @@ type (
 	// RPC contains the rpc config items
 	RPC struct {
 		// GRPCPort is the port  on which gRPC will listen
-		GRPCPort int `yaml:"grpcPort"`
+		GRPCPort      int `yaml:"grpcPort"`
+		boundGRPCPort *int
```
[the remark this answers is not in the document]
Comment (edmondop) [reply common/config/config.go:92]: Fixed with https://github.com/temporalio/temporal/pull/4713/commits/0cc91e4fbd4a61784b6611663d56ea073bbbf005

File: common/rpc/rpc.go [hunk 3]
```diff
@@ -148,6 +148,8 @@ func (d *RPCFactory) GetGRPCListener() net.Listener {
 		if err != nil {
 			d.logger.Fatal("Failed to start gRPC listener", tag.Error(err), tag.Service(d.serviceName), tag.Address(hostAddress))
 		}
+		d.logger.Info("Setting bound port for", tag.Service(d.serviceName))
```
[the remark this answers is not in the document]
Comment (edmondop) [reply common/rpc/rpc.go]: Fixed with https://github.com/temporalio/temporal/pull/4713/commits/0cc91e4fbd4a61784b6611663d56ea073bbbf005

File: common/rpc/test/rpc_common_test.go [hunk 4]
```diff
@@ -98,9 +98,13 @@ func startHelloWorldServer(s *suite.Suite, factory *TestFactory) (*grpc.Server,
 	server := grpc.NewServer(opts...)
 	greeter := &HelloServer{}
 	helloworld.RegisterGreeterServer(server, greeter)
-
+	previousValue := rpcTestCfgDefault.GetGRPCPort()
 	listener := factory.GetGRPCListener()
-
+	s.NotNil(rpcTestCfgDefault.GetGRPCPort())
```
[the remark this answers is not in the document]
Comment (edmondop) [reply common/rpc/test/rpc_common_test.go]: Great catch. It was something that I didn't cleanup after breaking down the PR in two changes . Fixed with https://github.com/temporalio/temporal/pull/4713/commits/0cc91e4fbd4a61784b6611663d56ea073bbbf005

**Comment by edmondop** [comment]
> When I change my local config to use port 0, I get errors when starting the server here: image image

Fixed through https://github.com/temporalio/temporal/pull/4713/commits/b598d50a4182bf688a26bee915433e0c88b05ada

File: common/membership/ringpop/factory.go [hunk 5]
```diff
@@ -251,19 +251,27 @@ func (factory *factory) getHostInfoProvider() (membership.HostInfoProvider, erro
...
 	if err != nil {
 		return nil, err
 	}
 
-	hostInfo := membership.NewHostInfoFromAddress(serviceAddress)
+	hostInfo := membership.NewHostInfoFromServiceHost(serviceAddress, serviceHost)
 	return membership.NewHostInfoProvider(hostInfo), nil
 }
+
+func extractHostFromAddress(address string) (string, error) {
```
Comment (underrun) [inline common/membership/ringpop/factory.go:271]: we could call this function from the `replaceServicePort` function in common/membership/ringpop/monitor.go - they both return the same error if `SplitHostPort` fails. alternately, `replaceSerivcePort` is only used in exactly one other case - maybe just use this new function and a `net.JoinHostPort`?

**Review by cretz — APPROVED** [review]
I don't think this is a needed feature, and I think frontends can't use it so it's only a half-measure, and the original issue misunderstands how port assignment works in OS's I think, and to really solve clashing in our CLI we need to be using random ports not offsets from a common random port.

But it's harmless I suppose.

**Comment by MichaelSnowden** [comment]
See https://github.com/temporalio/temporal/issues/4329#issuecomment-1870542739

**Pull request closed without merging**